### PR TITLE
Controller and route added for profile&availability #6

### PR DESCRIPTION
**Pull request by JAFAJ0**
### What this PR does (required):
- list PR functionality here
- Profile
- get(by userId)
- post(see userId, phone number, email)
- update
Availability
- get( a list with id)
- post( Cannot overlap time with the same id)
- update(get by availabilityId, Cannot overlap time with the same id)

### Any information needed to test this feature (required):
- Use tools like postsql

### CODE REVIEW

File: server/controllers/Availability.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
+const mongoose = require("mongoose");
+const Availability = require("../Models/Availability");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /Availability
+// @desc the Availability list  of the relevant user
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:7]: Good docummentation

File: server/controllers/Availability.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    const availability = await Availability.find({
+        $or: [{ userId: userId }]
+    });
+
+    res.send({ availability });
+});
+
+// @route POST /Availability
+// @desc Create a new Availability time solts
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:20]: :+1:

File: server/controllers/Availability.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+exports.createAvailability = asyncHandler(async (req, res, next) => {
+    const userId = req.user.id;
+
+    const { startDate, endDate } = req.body;
+
+    if (!startDate || !endDate) {
+        res.status(400);
+        throw new Error("Incomplete date field");
+    }
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:29]: A good security practice

File: server/controllers/Availability.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    const startDateExists = await Availability.findOne({ userId, startDate });
+
+    if (startDateExists) {
+        res.status(400);
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:33]: Correct status code

File: server/controllers/Availability.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        throw new Error("There is already a availability period with this start date");
+    }
+    const endDateExists = await Availability.findOne({ userId, endDate });
+
+    if (endDateExists) {
+        res.status(400);
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:39]: :+1:

File: server/controllers/Availability.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    };
+
+    const resp = await Availability.create(availability);
+
+    res.status(201).json(resp);
+});
+
+// @route PATCH /Availability
+// @desc Update Availability period with approved or decline
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:56]: :+1:

File: server/controllers/Availability.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    const endDateExists = await Availability.findOne({ userId, endDate });
+
+    if (endDateExists) {
+        res.status(400);
+        throw new Error("There is already a availability period with this end date");
+    }
+
+    await availability.save();
+
+    res.status(200).json(availability);
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:95]: Nice one!

File: server/controllers/Profile.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
+const mongoose = require("mongoose");
+const Profile = require("../Models/Profile");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /profile
+// @desc the profile of the relevant user
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:7]: :+1:

File: server/controllers/Profile.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    const userId = req.user.id;
+
+    const profile = await Profile.findOne({ userId });
+
+    res.send({ profile });
+});
+
+// @route POST /profile
+// @desc Create a new profile
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:18]: :+1:

File: server/controllers/Profile.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+exports.createProfile = asyncHandler(async (req, res, next) => {
+    const userId = req.user.id;
+
+    const { firstName, lastName, email, phoneNum } = req.body;
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:22]: You can first validate all these fields and return a bad request status if any of them is required but missing in the request body.

File: server/controllers/Profile.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    const emailExists = await Profile.findOne({ email });
+
+    if (emailExists) {
+        res.status(400);
+        throw new Error("The email is already in use");
+    }
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:29]: :+1:

File: server/controllers/Profile.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (phoneNumExists) {
+        res.status(400);
+        throw new Error("The phone number is already in use");
+    }
+
+    if (!firstName || !lastName || !email || !phoneNum) {
+        res.status(400);
+        throw new Error("Incomplete profile field");
+    }
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:41]: I think this should come first

File: server/controllers/Profile.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    const resp = await Profile.create(profile);
+
+    res.status(201).json(resp);
+});
+
+// @route PATCH /Profile
+// @desc Update profile with approved or decline
+// @access Private
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:58]: :+1:

File: server/controllers/Profile.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+exports.updateProfile = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:59]: You can remove `next` if not used

File: server/controllers/Profile.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+    const profile = await Profile.findOne({ userId });
+
+    res.send({ profile });
+});
+
+// @route POST /profile
+// @desc Create a new profile
+// @access Private
+exports.createProfile = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:19]: You can remove `next` if not used

File: server/controllers/Profile.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
+const mongoose = require("mongoose");
+const Profile = require("../Models/Profile");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /profile
+// @desc the profile of the relevant user
+// @access Private
+exports.getProfile = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/Profile.js:8]: You can remove `next` if not used

File: server/controllers/Availability.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+    const resp = await Availability.create(availability);
+
+    res.status(201).json(resp);
+});
+
+// @route PATCH /Availability
+// @desc Update Availability period with approved or decline
+// @access Private
+exports.updateAvailability = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:57]: You can remove `next` if not used

File: server/controllers/Availability.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
+const mongoose = require("mongoose");
+const Availability = require("../Models/Availability");
+const asyncHandler = require("express-async-handler");
+
+// @route GET /Availability
+// @desc the Availability list  of the relevant user
+// @access Private
+exports.getAvailability = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:8]: You can remove `next` if not used

File: server/controllers/Availability.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+        $or: [{ userId: userId }]
+    });
+
+    res.send({ availability });
+});
+
+// @route POST /Availability
+// @desc Create a new Availability time solts
+// @access Private
+exports.createAvailability = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/Availability.js:21]: You can remove `next` if not used

File: server/models/Availablity.js [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    required: true,
+    },
+    startDate: {
+        type: Date,
+        required: true
+    },
+    endDate: {
+        type: Date,
+        required: true
+    }
```
Comment (ezeilo-su) [inline server/models/Availablity.js:16]: :+1:

File: server/models/Profile.js [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+        default: ''
+    },
+    description: {
+        type: String,
+        default: ''
+    },
+    photoUrl: {
+        type: String,
+        default: ''
+    },
```
Comment (ezeilo-su) [inline server/models/Profile.js:41]: :+1:

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
What a neat code :+1: 
Just check out the few comments below and fix them.
Happy coding!

File: server/controllers/Availability.js [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+const asyncHandler = require("express-async-handler");
+
+// @route GET /Availability
+// @desc the Availability list  of the relevant user
+// @access Private
+exports.getAvailability = asyncHandler(async (req, res) => {
+    const userId = req.user.id;
+
+    const availability = await Availability.find({
+        $or: [{ userId: userId }]
```
Comment (rajivtitus) [inline server/controllers/Availability.js:12]:
```suggestion
         userId
```

I don't think we need an $or condition as we are only searching for the 'userId' field :)

File: server/controllers/Availability.js [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+exports.updateAvailability = asyncHandler(async (req, res) => {
+    const availabilityId = req.params.id;
+    if (!mongoose.Types.ObjectId.isValid(req.params.id)) {
+        throw new Error("Cannot find such available period");
+    }
+    const userId = req.user.id;
+
+    const { availabilityupdates } = req.body;
+
+    const availability = await Availability.findById(availabilityId);
```
Comment (rajivtitus) [inline server/controllers/Availability.js:66]: Optional: Only the profile owner should be allowed to update their availability. The way we currently have it setup, any user can modify any other user's availability :)

File: server/controllers/Profile.js [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    res.status(201).json(resp);
+});
+
+// @route PATCH /Profile
+// @desc Update profile with approved or decline
+// @access Private
+exports.updateProfile = asyncHandler(async (req, res) => {
+    const userId = req.user.id;
+
+    const { profileupdates } = req.body;
```
Comment (rajivtitus) [inline server/controllers/Profile.js:54]:
```suggestion
    const { profileUpdates } = req.body;

```

camelCase :)

File: server/controllers/Profile.js [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+
+    const profile = await Profile.findById(userId);
+
+    if (!profile) {
+        res.status(404);
+        throw new Error("User id do not exist! Error happened!");
+    }
+
+    profile.status = profileupdates;
```
Comment (rajivtitus) [inline server/controllers/Profile.js:63]:
```suggestion
    profile.status = profileUpdates;

```

File: server/routes/Profile.js [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+    getProfile,
+    createProfile,
+    updateProfile,
+} = require("../controllers/Profile");
+
+router.route("/").get(protect, getProfile);
+
+router.route("/").post(protect, createProfile);
+
+router.route("/:id").patch(protect, updateProfile);
```
Comment (rajivtitus) [inline server/routes/Profile.js:14]:
```suggestion
router.route("/").patch(protect, updateProfile);

```

I don't see the path params being using in the 'updateProfile' controller

**Review by rajivtitus — APPROVED** [review]
Good work with this! Just some minor suggestions, please address them before merging. Thanks!

**Pull request merged**